src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/HotReload/ProjectHotReloadSessionManager.cs

```diff
@@ -0,0 +1,233 @@
+                    ImmutableInterlocked.TryAdd(ref _activeSessions, processId, _pendingSessionState);
+                }
+
+                _pendingSessionState = null;
+            }
+        }
+
+        public async Task<bool> TryCreatePendingSessionAsync(IDictionary<string, string> environmentVariables)
+        {
+            if (await DebugFrameworkSupportsHotReloadAsync()
+                && await GetDebugFrameworkVersionAsync() is string frameworkVersion)
+            {
+                string name = $"{Path.GetFileNameWithoutExtension(_project.FullPath)}:{_nextUniqueId++}";
+                HotReloadState state = new(this);
+                IProjectHotReloadSession? projectHotReloadSession = _projectHotReloadAgent.Value.CreateHotReloadSession(name, frameworkVersion, state);
+
+                if (projectHotReloadSession is not null)
+                {
+                    state.Session = projectHotReloadSession;
+                    await projectHotReloadSession.ApplyLaunchVariablesAsync(environmentVariables, default);
+                    _pendingSessionState = state;
+
+                    return true;
+                }
+            }
+
+            _pendingSessionState = null;
+            return false;
+        }
+
+        private async Task<bool> DebugFrameworkSupportsHotReloadAsync()
+        {
+            ConfiguredProject? configuredProjectForDebug = await GetConfiguredProjectForDebugAsync();
+            if (configuredProjectForDebug is null)
+            {
+                return false;
+            }
+
+            return configuredProjectForDebug.Capabilities.AppliesTo("SupportsHotReload");
+        }
+
+        private Task<ConfiguredProject?> GetConfiguredProjectForDebugAsync() =>
+            _activeDebugFrameworkServices.GetConfiguredProjectForActiveFrameworkAsync();
+
+        private async Task<string?> GetDebugFrameworkVersionAsync()
+        {
+            ConfiguredProject? configuredProjectForDebug = await GetConfiguredProjectForDebugAsync();
+            if (configuredProjectForDebug is null)
+            {
+                return null;
+            }
+
+            Assumes.Present(configuredProjectForDebug.Services.ProjectPropertiesProvider);
+            IProjectProperties commonProperties = configuredProjectForDebug.Services.ProjectPropertiesProvider.GetCommonProperties();
+            string targetFrameworkVersion = await commonProperties.GetEvaluatedPropertyValueAsync(ConfigurationGeneral.TargetFrameworkVersionProperty);
+
+            if (targetFrameworkVersion.StartsWith("v", StringComparison.OrdinalIgnoreCase))
+            {
+                targetFrameworkVersion = targetFrameworkVersion.Substring(startIndex: 1);
+            }
+
+            return targetFrameworkVersion;
+        }
+
+        private async Task ProcessExitedAsync(HotReloadState hotReloadState)
+        {
+            Assumes.NotNull(hotReloadState.Session);
+            Assumes.NotNull(hotReloadState.Process);
```

Comment:
It might be hard to know which of this caused an exception, if we don't have line numbers. Perhaps use `Requires.NotNull`, passing the property name instead of the parameter name. Ditto below in `StopProjectAsync`.

Reply:
If we've got symbols, we should have line numbers.